### PR TITLE
Featurebranch dwk1265 entity definition and storage

### PR DESCRIPTION
**Pull request by JanPeterDatakind**
Entities are now created as basic SQL statements while the type of materialization is set to "view" per default in the dbt_project.yml file. 
Also, Appsmith now contains a preview of the entity creation, showing a table of the dbt model (entity)

image

### CODE REVIEW

File: db/dot/4-upload_sample_dot_data.sql [hunk 1]
```diff
@@ -6,30 +6,20 @@ INSERT INTO dot.entity_categories VALUES('ZAG', 'Zagreb airport flights');
...
-INSERT INTO dot.configured_entities VALUES('ScanProject1', 'all_airports_data', 'ALL', '{{ config(materialized=''view'') }}
-{% set schema = <schema> %}
-select *
-from {{ schema }}.airport_data   ','2021-12-07 00:00:00+00','2021-12-07 00:00:00+00','Matt');
-
-INSERT INTO dot.configured_entities VALUES('ScanProject1', 'airlines_data', 'ALL', '{{ config(materialized=''view'') }}
-{% set schema = <schema> %}
-select DISTINCT airline
-from {{ schema }}.flight_data   ','2021-12-07 00:00:00+00','2021-12-07 00:00:00+00','Matt');
+INSERT INTO dot.configured_entities VALUES('ScanProject1', 'all_flight_data', 'ALL', 'select *
```
Comment (lrnzcig) [inline db/dot/4-upload_sample_dot_data.sql:9]: would not this remove the possibility to have a different schema name for the self tests?
I don't think we really _need_ to keep the possibility, it is just we should maybe add a statement in the README file

File: dot/self_tests/data/expected/dot_model__all_flight_data.sql [hunk 2]
```diff
@@ -1,4 +1,2 @@
-{{ config(materialized='view') }}
-{% set schema = 'schema_project' %}
 select *
-from {{ schema }}.flight_data 
+from self_tests_public.flight_data
```
Comment (lrnzcig) [inline dot/self_tests/data/expected/dot_model__all_flight_data.sql:2]: same as above

**Review by lrnzcig — COMMENTED** [review]
looks fine to me

I understand from the summary of the PR that you want to simplify the definition of dbt entities, so it is more that fine that we loose the possibility of having a different name for the selftest schemas... I will add a comment in the README in another PR if you want

I "comment" instead of "approve" since don't have an opinion of your appsmith changes -don't know much about the tool and the json is unreadable

thanks

**Comment by JanPeterDatakind** [comment]
@lrnzcig thanks for the review. Sure, it would be great if you could add a comment about the fixed schema to the ReadMe in your PR! About the Appsmith issue: I checked with @dividor today and will explore Appsmith's version control via GitHub so it will be easier to track and review changes moving forward.

File: dot/self_tests/data/expected/dot_model__all_flight_data.sql (hunk 2 above)
Comment (dividor) [reply dot/self_tests/data/expected/dot_model__all_flight_data.sql:2]: Hi @JanPeterDatakind and @lrnzcig 

The two requirements are equally valid, to have something easy for user to maintain with a data preview, plus the ability for self-tests. The schema is determined by project definition, the user shouldn't really be allowed to override that or know about it in their SQL statement.

How about something like ...

1. We need a tooltip to guide the user to specify their SQL statement assuming *no* schema. Also make it the placeholder text in the SQL box (Appsmith supports this)
2. When saving the query to the DB, we add the "{{ schema }}." macro ahead of any tables/views in the query
3. When retrieving the query from the DB for display in Appsmith we remove "{{ schema }}"  on-the-fly
4. For the query used to get the preview in the UI we replace on-the-fly {{ schema }}. with the schema from the project table
5. Though we will tackle more comprehensively in another PR, there is a race condition where a project is set up but airflow has not been run. In this scenario the data schema will not exist and airflow needs to run one time. Let's add a hard stop with pop up on the entity screen which prevents the user maintaining the entity if the project schema does not exist
6. We add a check to the SQL, if the user has accidentally provided a schema in their query, detect it and ask they remove

In the above, 1 will be a bit confusing, but I don't see a way around it without exposing Jinja macros, which users will mess up. The above will also be challenging as parsing table names is not always easy from a SQL statement (LLMs do a great job though). I say let's start with this and refine as issues arise.

How does the above sound?

Comment (lrnzcig) [reply dot/self_tests/data/expected/dot_model__all_flight_data.sql:2]: @dividor I think it is simpler than that

`dot/self_tests/data/expected/dot_model__all_flight_data.sql` is just a expected result for a self test, we don't need to make it configurable if we assume that the schema of the output of the selftests is not configurable but always `self_tests_public`

afaict we don't need the other config either; we just need to enforce the schema names in the selftests

makes sense?

**Comment by lrnzcig** [comment]
@JanPeterDatakind added comments to the README in the commit just above

**Pull request closed without merging**